File: src/modules/admin/services/admin.service.ts
import {
  ConflictException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import * as bcrypt from 'bcrypt';
import { Repository } from 'typeorm';
import { Department } from '../../department/entities/department.entity';
import { Employee } from '../../employee/entities/employee.entity';
import { AccountStatus } from '../../employee/enums/account-status.enum';
import { EmployeeRole } from '../../employee/enums/employee-role.enum';
import {
  CreateAdminUserDto,
  ResetPasswordDto,
  UpdateAdminUserDto,
  UpdateUserRoleDto,
  UpdateUserStatusDto,
} from '../dto/admin.dto';

@Injectable()
export class AdminService {
  private readonly saltRounds = 10;

  constructor(
    @InjectRepository(Employee)
    private readonly employeeRepository: Repository<Employee>,
    @InjectRepository(Department)
    private readonly departmentRepository: Repository<Department>,
  ) {}

  findAllUsers(): Promise<Employee[]> {
    return this.employeeRepository.find({
      relations: { department: true },
      order: { createdAt: 'ASC' },
    });
  }

  async findUser(id: string): Promise<Employee> {
    const user = await this.employeeRepository.findOne({
      where: { id },
      relations: { department: true },
    });
    if (!user) {
      throw new NotFoundException(`User with id "${id}" not found`);
    }
    return user;
  }

  async createUser(dto: CreateAdminUserDto): Promise<Employee> {
    const existing = await this.employeeRepository.findOne({
      where: { email: dto.email },
    });
    if (existing) {
      throw new ConflictException('Email already registered');
    }

    let department: Department | undefined;
    if (dto.departmentId) {
      department =
        (await this.departmentRepository.findOne({
          where: { id: dto.departmentId },
        })) ?? undefined;
      if (!department) {
        throw new NotFoundException('Department not found');
      }
    }

    const passwordHash = await bcrypt.hash(dto.password, this.saltRounds);
    const user = this.employeeRepository.create({
      firstName: dto.firstName,
      lastName: dto.lastName,
      email: dto.email,
      password: passwordHash,
      role: dto.role ?? EmployeeRole.EMPLOYEE,
      accountStatus: dto.accountStatus ?? AccountStatus.ACTIVE,
      department,
    });

    return this.employeeRepository.save(user);
  }

  async updateUser(id: string, dto: UpdateAdminUserDto): Promise<Employee> {
    const user = await this.findUser(id);
    if (dto.firstName) user.firstName = dto.firstName;
    if (dto.lastName) user.lastName = dto.lastName;
    if (dto.email) user.email = dto.email;
    if (dto.departmentId !== undefined) {
      if (dto.departmentId === null) {
        user.department = undefined;
      } else {
        const department = await this.departmentRepository.findOne({
          where: { id: dto.departmentId },
        });
        if (!department) {
          throw new NotFoundException('Department not found');
        }
        user.department = department;
      }
    }
    return this.employeeRepository.save(user);
  }

  async updateRole(id: string, dto: UpdateUserRoleDto): Promise<Employee> {
    const user = await this.findUser(id);
    user.role = dto.role;
    return this.employeeRepository.save(user);
  }

  async updateStatus(id: string, dto: UpdateUserStatusDto): Promise<Employee> {
    const user = await this.findUser(id);
    user.accountStatus = dto.accountStatus;
    return this.employeeRepository.save(user);
  }

  async resetPassword(id: string, dto: ResetPasswordDto): Promise<void> {
    const user = await this.findUser(id);
    user.password = await bcrypt.hash(dto.password, this.saltRounds);
    await this.employeeRepository.save(user);
  }

  async removeUser(id: string): Promise<void> {
    const user = await this.findUser(id);
    await this.employeeRepository.remove(user);
  }
}

File: src/EquipmentApi/Services/AdminService.cs
using EquipmentApi.Data;
using EquipmentApi.Dtos;
using EquipmentApi.Entities;
using EquipmentApi.Enums;
using EquipmentApi.Exceptions;
using Microsoft.EntityFrameworkCore;

namespace EquipmentApi.Services;

public class AdminService
{
    private const int SaltRounds = 10;

    private readonly AppDbContext _db;
    private readonly AuthService _authService;

    public AdminService(AppDbContext db, AuthService authService)
    {
        _db = db;
        _authService = authService;
    }

    public Task<List<Employee>> FindAllUsersAsync(CancellationToken cancellationToken = default)
    {
        return _db.Employees
            .Include(user => user.Department)
            .OrderBy(user => user.CreatedAt)
            .ToListAsync(cancellationToken);
    }

    public async Task<Employee> FindUserAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var user = await _db.Employees
            .Include(e => e.Department)
            .FirstOrDefaultAsync(e => e.Id == id, cancellationToken);

        if (user is null)
        {
            throw new NotFoundException($"User with id \"{id}\" not found");
        }

        return user;
    }

    public async Task<Employee> CreateUserAsync(
        CreateAdminUserDto dto,
        CancellationToken cancellationToken = default)
    {
        var existing = await _db.Employees.AnyAsync(e => e.Email == dto.Email, cancellationToken);
        if (existing)
        {
            throw new ConflictException("Email already registered");
        }

        Department? department = null;
        if (dto.DepartmentId is not null)
        {
            department = await _db.Departments
                .FirstOrDefaultAsync(d => d.Id == dto.DepartmentId, cancellationToken);
            if (department is null)
            {
                throw new NotFoundException("Department not found");
            }
        }

        var user = new Employee
        {
            FirstName = dto.FirstName,
            LastName = dto.LastName,
            Email = dto.Email,
            Password = BCrypt.Net.BCrypt.HashPassword(dto.Password, SaltRounds),
            Role = dto.Role ?? EmployeeRole.Employee,
            AccountStatus = dto.AccountStatus ?? AccountStatus.Active,
            DepartmentId = department?.Id,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow,
        };

        _db.Employees.Add(user);
        await _db.SaveChangesAsync(cancellationToken);
        user.Department = department;
        return user;
    }

    public async Task<Employee> UpdateUserAsync(
        Guid id,
        UpdateAdminUserDto dto,
        CancellationToken cancellationToken = default)
    {
        var user = await FindUserAsync(id, cancellationToken);

        if (dto.FirstName is not null)
        {
            user.FirstName = dto.FirstName;
        }

        if (dto.LastName is not null)
        {
            user.LastName = dto.LastName;
        }

        if (dto.Email is not null)
        {
            user.Email = dto.Email;
        }

        if (dto.DepartmentId.HasValue)
        {
            if (dto.DepartmentId.Value == Guid.Empty)
            {
                user.DepartmentId = null;
                user.Department = null;
            }
            else
            {
                var department = await _db.Departments
                    .FirstOrDefaultAsync(d => d.Id == dto.DepartmentId, cancellationToken);
                if (department is null)
                {
                    throw new NotFoundException("Department not found");
                }

                user.DepartmentId = department.Id;
                user.Department = department;
            }
        }

        await _db.SaveChangesAsync(cancellationToken);
        return user;
    }

    public async Task<Employee> UpdateRoleAsync(
        Guid id,
        UpdateUserRoleDto dto,
        CancellationToken cancellationToken = default)
    {
        var user = await FindUserAsync(id, cancellationToken);
        user.Role = dto.Role;
        await _db.SaveChangesAsync(cancellationToken);
        return user;
    }

    public async Task<Employee> UpdateStatusAsync(
        Guid id,
        UpdateUserStatusDto dto,
        CancellationToken cancellationToken = default)
    {
        var user = await FindUserAsync(id, cancellationToken);
        user.AccountStatus = dto.AccountStatus;
        await _db.SaveChangesAsync(cancellationToken);

        if (dto.AccountStatus == AccountStatus.Inactive)
        {
            await _authService.RevokeAllSessionsAsync(id, cancellationToken);
        }

        return user;
    }

    public async Task ResetPasswordAsync(
        Guid id,
        ResetPasswordDto dto,
        CancellationToken cancellationToken = default)
    {
        var user = await FindUserAsync(id, cancellationToken);
        user.Password = BCrypt.Net.BCrypt.HashPassword(dto.Password, SaltRounds);
        await _db.SaveChangesAsync(cancellationToken);
        await _authService.RevokeAllSessionsAsync(id, cancellationToken);
    }

    public async Task RemoveUserAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var user = await FindUserAsync(id, cancellationToken);
        _db.Employees.Remove(user);
        await _db.SaveChangesAsync(cancellationToken);
    }
}
